### ISSUE TITLE
Unify parametrization of circuit sizes

### ISSUE THREAD

**Issue by ed255** [privacy-scaling-explorations/zkevm-circuits#954]
See https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/650#issuecomment-1319441372

Our current circuits can be parametrized with a maximum number of operations supported:
- TxCircuit: max number of txs
- CopyCircuit: max number of copied bytes
- StateCircuit: max number of rw entries
- ExpCircuit: ?
- EvmCircuit: max number of rows (Evm Steps take a dynamic number of rows)
- MPTCircuit: max number of MPT proofs?
- PiCircuit: max number of txs + max calldata bytes
- BytecodeCircuit: max number of bytecode bytes

Sometimes these numbers don't correspond to the number of rows enabled (because an operation takes several rows)

For unit tests we'd like the number to be chosen at runtime based on the test input.
- For this, the bytecode circuit is using `Option<usize>` to specify the number of rows, and when set to `None`, it calculates the number of rows based on the input
- For other circuits it's difficult to decide this value based on the input because there are dependencies (Evm depends on max_txs and max_rws), but maybe it can be done

Tasks:
- Explore if we could set all the parameters as `Option<usize>` so that they can be optionally set, or calculated based on inputs
- Unify the meaning of the size parameter
    - I think what makes more sense is setting the number of operations that the circuit supports.  Note that a check should be done to make sure that the given number is possible to achieve based on the `k` value and the `unusable_rows`.
- Put all these parameters in `CircuitParams`, so that they can all be defined in a single place.

**Comment by ed255** [comment]
There's some uncertainty on the scope of this issue, because:
- Some of the tasks have already been done
- Not all circuits support dynamically choosing the parameter, and we may want to work on it or not.

**Comment by ChihChengLiang** [comment]
@davidnevadoc There seems to be no associated PR for this task. Are you working on a branch locally?

**Comment by ChihChengLiang** [comment]
I agree with the unifying the parameters part. We have integrated all parameters into `CircuitParams`.

But it looks like we haven't got the dynamic parameter selection yet. We are still manually assigning the values of the parameters and using default magic numbers.

 https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4669a87cd3ce4e25e9e08d3e5a51306c60858cb5/zkevm-circuits/src/evm_circuit/execution/sha3.rs#L172-L175 

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4669a87cd3ce4e25e9e08d3e5a51306c60858cb5/bus-mapping/src/circuit_input_builder.rs#L73-L87

Let's talk about this in the next call.

**Comment by ChihChengLiang** [comment]
Pending review of the issue from @ed255

**Comment by ed255** [comment]
I think there's work that can be done towards having more circuits that support runtime size to have less hard-coded numbers and possibly faster unit test execution.
The current "standard" is that circuits that support runtime number of rows need to receive a `max_foo = 0` value to only enable as many rows as the input requires.
Here are the circuits that support this:
- keccak
Here are the circuits that don't:
- bytecode: This one should be easy to do
- copy: easy to do
- evm: Uses `max_rws` but there's a dependency with state circuit (should use the same `max_rws`), I think it also uses max_txs which must match with tx circuit
- exp: Should be easy to do.  Also this has the constructor `ExpCircuit::new(exp_events: Vec<ExpEvent>, max_exp_rows: usize)` but should be renamed to `max_exp_steps`.  `max_exp_rows = max_exp_steps * OFFSET_INCREMENT`
- pi: Uses `max_txs` and `max_calldata` which must match with tx circuit
- state: Uses `max_rws` which must match with evm circuit
- tx: uses `max_txs` and `max_calldata` which must match with pi and evm circuit

After writing this I realize there are two possible approaches:
a. Each circuit can take a size of 0, and dynamically choose based on input
b. Circuits don't support dynamic sizes, but instead we have a method that takes a Block and returns a CircuitsParams adjusted for that block.

I think option `b` would be a nice alternative.  Nevertheless there's a challenge implementing `b` and it's that we have a circular dependency:
- The CircuitInputBuilder requires `max_rws` to produce the `EndBlock` step: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/0e1d230191d3a1b3ea4b27034c73a545a6e66158/bus-mapping/src/circuit_input_builder.rs#L211
- Calculating CircuitsParams dynamically requires knowing the output of the CircuitInputBuilder
In my opinion this circular dependency is ugly (I say this even though I introduced it), because it couples the circuit inputs generation with the circuits parameter selection.  I think ideally they should be independent, so that we can generate the inputs, and then pick the circuits parameters.  I also believe that there are different ways to resolve this!  Example
  - Calling `set_end_block` with a dummy `max_rws` first, then picking the parameters, and then adjusting the circuit inputs
  - Implementing the `set_end_block` in a struct that is closer to the circuits, so that it is called after picking the parameters

So in summary I think there are still some improvements to be done, and I believe the approach `b` will have bigger benefits.

**Comment by ChihChengLiang** [comment]
@ed255 ping @davidnevadoc on how to proceed this.
